component/discovery/file/file.go

```diff
@@ -0,0 +1,346 @@
+}
+
+// UnmarshalRiver implements river.Unmarshaler.
+func (a *Arguments) UnmarshalRiver(f func(interface{}) error) error {
+	*a = getDefault()
+	type arguments Arguments
+	return f((*arguments)(a))
+}
+
+// Update satisfies the component interface.
+func (c *Component) Update(args component.Arguments) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	c.args = args.(Arguments)
+	c.reconcileWatchesWithWatcher()
+	return nil
+}
+
+// Run satisfies the component interface.
+func (c *Component) Run(ctx context.Context) error {
+	watchDog := time.NewTicker(c.args.UpdatePeriod)
+	timerDuration := c.args.UpdatePeriod
+	update := func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+
+		// See if there is anything new we need to check.
+		c.reconcileWatchesWithWatcher()
+		// Update the exports with the targets. Should only be called if changes occurred.
+		c.checkOnStateChanged()
+		// Check to see if our ticker timer needs to be reset.
+		if timerDuration != c.args.UpdatePeriod {
+			watchDog.Reset(c.args.UpdatePeriod)
+			timerDuration = c.args.UpdatePeriod
+		}
+	}
+	// Trigger initial check
+	update()
+	defer watchDog.Stop()
+	for {
+		select {
+		case fe := <-c.watcher.Events:
+			c.fsnotifyTrigger(fe)
+		case err := <-c.watcher.Errors:
+			level.Error(c.opts.Logger).Log("msg", "error with fsnotify", "err", err)
+		case <-watchDog.C:
+			// This triggers a check for any new paths, along with pushing new targets.
+			update()
+		case <-ctx.Done():
+			c.mut.Lock()
+			err := c.watcher.Close()
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error closing watcher", "err", err)
+			}
+			c.mut.Unlock()
+			return nil
+		}
+	}
+}
+
+// reconcileWatchesWithWatcher checks for any new directories that have been added along with verifying
+// that the args and watchers are in sync.
+func (c *Component) reconcileWatchesWithWatcher() {
+	includedPaths := c.args.getPaths()
+	excludedPaths := c.args.getExcluded()
+	expandedPaths, err := getPaths(includedPaths)
+	if err != nil {
+		level.Error(c.opts.Logger).Log("msg", "error expanding paths", "err", err)
+		return
+	}
+	alreadyWatching := c.watcher.WatchList()
+	alreadyWatchingDir := make(map[string]struct{})
+	for _, p := range alreadyWatching {
+		alreadyWatchingDir[p] = struct{}{}
+	}
+	// Ensure all the paths are added.
+	for _, n := range expandedPaths {
+		fi, fileErr := os.Stat(n)
+		if fileErr != nil {
+			level.Error(c.opts.Logger).Log("msg", "error getting os stats", "err", err)
+			continue
+		}
+		if fi.IsDir() {
+			// Check to see if we are already watching.
+			if _, found := alreadyWatchingDir[n]; found {
+				continue
+			}
+			err = c.watcher.Add(n)
+			if err != nil {
+				level.Error(c.opts.Logger).Log("msg", "error adding path to watcher", "err", err)
+			}
+		} else {
+			exclude := false
+			for _, excluded := range excludedPaths {
+				if match, _ := doublestar.Match(excluded, n); match {
+					exclude = true
+					break
+				}
+			}
+			if exclude {
+				continue
+			}
```

Comment:
I think maybe this snippet might belong inside of the `getPaths` helper. This way we'd get what we want for `expandedPaths` from the start.

As a plus, I think these would never appear in the list of watchedFiles so we might simplify the similar check that appears below.

Reply:
I'm not sure I understood what you meant here.

I feel the logic here doesn't exactly match what Promtail did and what we want to do here.

In the current code, we gather all the paths, glob expand them, and then drop them if _any_ of the targets happened to match the excludion.

I believe what we'd like to do is treat each target _separately_, so each `__path__` and possibly `__path_exclude__` that we encounter is globbed and exported with its own set of labels.